### ISSUE TITLE
Difference of predicted masks between .pb model and .onnx model which is converted from the former one.

### ISSUE THREAD

**Issue by SpongeYao** [onnx/tensorflow-onnx#812]
**Describe the bug**
I have converted maskRCNN .pb file to .onnx file by tf2onnx.convert without any error message. The message is shown as follow:
![image](https://user-images.githubusercontent.com/26354700/75322492-cff8d880-58ad-11ea-8ae0-cc0d59462cc5.png)

Then I  individually used .pb file and .onnx file to predict the object with the same image. However, the predicted mask(33x33) of .onnx file is different from the one from .pb file, shown as follow. 
_**predicted mask of .pb file:**_
![mask_tfmodel](https://user-images.githubusercontent.com/26354700/75323025-07b45000-58af-11ea-9efc-292f9627b4b9.PNG)

_**predicted mask of .onnx file:**_
![mask_onnx](https://user-images.githubusercontent.com/26354700/75323054-100c8b00-58af-11ea-94b9-4352d13a1cfe.PNG)
If we look carefully, the difference of each pixel value between predicted mask of .pb file and of .onnx file is quite huge. The difference could be 1. (The range of pixel value in predicted mask is [0,1] )

However, the predicted bounding box and the scores of each object are almost exactly the same.
_**predicted bounding box (Left: .pb file ; Right: .onnxfile):**_ 
![roi_Ltf_Ronnx](https://user-images.githubusercontent.com/26354700/75323259-7d202080-58af-11ea-9274-8a42eff6f84b.PNG)

**_score of each object(Left: .pb file ; Right: .onnxfile):_**
![score_Ltf_Ronnx](https://user-images.githubusercontent.com/26354700/75323309-93c67780-58af-11ea-9e49-74d72cbcc884.PNG)

**System information**
- OS: Windows 10
- Tensorflow Version: 1.12.0
- Python Version: 3.6.5
- onnx Version: 1.6.0
- onnxruntime-gpu Version: 1.1.2

**To Reproduce**
Describe steps/code to reproduce the behavior:
The MaskRCNN architecture is from [Tensorflow object detection](https://github.com/tensorflow/models/tree/master/research/object_detection). The frozen_inference_graph.pb file is generated by Tensorflow object detection directly after training. 

Then I use the following command to convert it to .onnx file:
`python -m tf2onnx.convert --graphdef frozen_inference_graph.pb --output frozen.onnx  --opset 11 --inputs image_tensor:0 --outputs  num_detections:0,detection_boxes:0,detection_scores:0,detection_classes:0,detection_masks:0`

Here is the message after the command is completed:
![run_tf2onnxCvt](https://user-images.githubusercontent.com/26354700/75324408-d721e580-58b1-11ea-81f6-91a5421f300b.PNG)

After creating .onnx I use the following command to get the prediction of .onnx file: 
```
import onnxruntime as rt
sess = rt.InferenceSession("frozen.onnx")
result = sess.run(outputs, {"image_tensor:0": image})
```

**Expected behavior**
Ideally, after converting .pb file to .onnx file, the prediction from both of these two file should be exactly the same. Currently, the predicted bounding box and score are the same; while, the predicted masks should have been exactly the same. 

Thanks for any advice in advance :)

**Comment by p890040** [comment]
Exactly same problem as mine.
Hopelly someone can give a help :)

**Comment by RandySheriffH** [comment]
@SpongeYao : could u please pinpoint the model referred from the link in case I am looking at the wrong one?

**Comment by SpongeYao** [comment]
> @SpongeYao : could u please pinpoint the model referred from the link in case I am looking at the wrong one?

@RandySheriffH  No problem
The model I used is _mask_rcnn_resnet101_atrous_coco_  from [Tensorflow detection model zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) as shown in below image:
![image](https://user-images.githubusercontent.com/26354700/75738355-08306900-5d3d-11ea-942c-6f841aee1c78.png)

If there is any extra information you need, just let me know. Thank you

**Comment by RandySheriffH** [comment]
@SpongeYao: 
Done some investigation - the culprit is CropAndResize op, in onnx we don't have an op fit in well. So instead we assemble one with a loop - 
https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/onnx_opset/nn.py
With image I have at hand, maximum abs-diff is below 0.001.
Would u mind share me your image for more tests?

**Comment by chiehpower** [comment]
> @SpongeYao:
> Done some investigation - the culprit is CropAndResize op, in onnx we don't have an op fit in well. So instead we assemble one with a loop -
> https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/onnx_opset/nn.py
> With image I have at hand, maximum abs-diff is below 0.001.
> Would u mind share me your image for more tests?

Hi @RandySheriffH, 
As you mentioned, if we meet this problem of using CropAndResize operation, how can we convert it properly? 
Because I knew that even the opset 11, it cannot still support this operation. 

Recently I also noticed this issue that I want to convert the .pb which is converted by Keras .h5 of Matterport MRCNN to onnx. 
After I used the summarize_graph to analyze, we can see that it used 8 CropAndResize. 

`
Op types used: 1790 Const, 693 Identity, 489 Reshape, 359 Mul, 276 Add, 137 BiasAdd, 134 Conv2D, 132 Sub, 112 Relu, 110 Rsqrt, 105 StridedSlice, 28 Pack, 28 Shape, 27 Transpose, 21 GatherV2, 16 StopGradient, 14 ConcatV2, 13 Cast, 12 RealDiv, 12 Where, 12 Minimum, 11 Enter, 11 Maximum, 10 GatherNd, 9 Equal, 9 Exp,` **8 CropAndResize** `, 7 Split, 6 ExpandDims, 5 Range, 5 Sum, 5 Max, 4 TopKV2, 4 Squeeze, 4 Sqrt, 4 Log, 3 NextIteration, 3 Pad, 3 Placeholder, 3 Switch, 3 Merge, 3 ResizeNearestNeighbor, 2 NonMaxSuppressionV3, 2 DenseToDenseSetOperation, 2 MatMul, 2 SparseToDense, 2 TensorArrayV3, 2 MaxPool, 2 Round, 2 Less, 2 Greater, 1 TensorArrayScatterV3, 1 TensorArraySizeV3, 1 Exit, 1 TensorArrayWriteV3, 1 Conv2DBackpropInput, 1 Unique, 1 ArgMax, 1 TensorArrayReadV3, 1 TensorArrayGatherV3, 1 Softmax, 1 Sigmoid, 1 PadV2, 1 GreaterEqual, 1 LoopCond, 1 LogicalAnd
`

Can you suggest us some ideas if we meet this kind of issue?

I will be grateful for any help you can provide.
Thank you.

**Comment by RandySheriffH** [comment]
@chiehpower : Like being said - the diff should be minor.
But tests show that it really affect the precision of your model, may be we could go with ort contrib op - "cropandresize" during conversion instead of assembly logic manually.

**Comment by SpongeYao** [comment]
> @SpongeYao:
> Done some investigation - the culprit is CropAndResize op, in onnx we don't have an op fit in well. So instead we assemble one with a loop -
> https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/onnx_opset/nn.py
> With image I have at hand, maximum abs-diff is below 0.001.
> Would u mind share me your image for more tests?

@RandySheriffH  No Problem.
[Here](https://drive.google.com/drive/folders/17zTuBv3EsfTG-2k1Tfl4qgu3qdxs4Bgt) is the link of the required files. And the files are as follow:
1. **frozen_inference_graph.pb**: the .pb file which is created by tensorflow api
2. **pipeline.config**: one of the setting files of Tensorflow object detection 
3. **temp.config**: one of the setting files of Tensorflow object detection
4. **frozen.onnx**: the onnx model file which is converted from frozen_inference_graph.pb
4. **image.szip**:  the zip file which contains the testing images.

After detecting the testing image inside images.zip, the difference of the detecting result between the frozen_inference_graph.pb file and frozen.onnx file are large. Hopefully, it will help your testing.

After you complete downloading the data, please let me know. Thank you:)

**Comment by RandySheriffH** [comment]
@SpongeYao : Thanks for sharing the images.
With merged PR
https://github.com/onnx/tensorflow-onnx/pull/830
tf2onnx now could utilize a contrib version of "cropandresize" with "--extra_opset com.microsoft:1"
However, the option still does not eliminate the diff oberved when using the assembled version - by image provide I could still see diff around 0.98.
Talked to @jiafatom, implementer of the contrib cropandresize, seems we have a reason to see a diff there. 
@jiafatom , would u mind to share some background please?

**Comment by jiafatom** [comment]
The implementation of contrib CropAndResize is a little different from tensorflow version, that is because we find tf version is not well handling some corner cases. So if we see difference from CropAndResize output, that is understandable. 

We have a keras version mask-rcnn, converted by [keras2onnx](https://github.com/onnx/keras-onnx), which also contains this CropAndResize. On output mask or bounding box, we don't see difference while testing many images. So from our test, we may see difference from CropAndResize, generally we don't see it affects final results.

**Comment by RandySheriffH** [comment]
Closing the issue but please feel free to reopen if further assistance required.
